snapcraft/elf/_patcher.py

```diff
@@ -0,0 +1,150 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2016-2022 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Use patchelf to patch ELF files."""
+
+import os
+import shutil
+import subprocess
+import tempfile
+from pathlib import Path
+from typing import List, Set
+
+from craft_cli import emit
+
+from snapcraft import utils
+
+from ._elf_file import ElfFile
+from .errors import PatcherError
+
+
+class Patcher:
+    """Hold the necessary logic to patch elf files."""
+
+    def __init__(
+        self, *, dynamic_linker: str, root_path: str, preferred_patchelf=None
+    ) -> None:
+        """Create a Patcher instance.
+
+        :param dynamic_linker: The path to the dynamic linker to set the ELF file to.
+        :param root_path: The base path for the snap being processed.
+        :param preferred_patchelf: patch the necessary elf_files with this patchelf.
+        """
+        self._dynamic_linker = dynamic_linker
+        self._root_path = root_path
+        self._patchelf_cmd = preferred_patchelf or utils.get_snap_tool("patchelf")
+        self._strip_cmd = utils.get_snap_tool("strip")
+
+    def patch(self, *, elf_file: ElfFile) -> None:
+        """Patch elf_file with the Patcher instance configuration.
+
+        If the ELF is executable, patch it to use the configured linker.
+        If the ELF has dependencies (DT_NEEDED), set an rpath to them.
+
+        :param elf_file: a data object representing an elf file and its attributes.
+
+        :raises PatcherError: if the ELF file cannot be patched.
+        """
+        patchelf_args = []
+        if elf_file.interp:
+            patchelf_args.extend(["--set-interpreter", self._dynamic_linker])
```

Comment:
we should only set the proposed interpreter if it differs from the current one.

Reply:
Comparing before setting.